feat(ui): progress interface

comfysage opened the pull request:
create an initial progress model as described in #32537 .

based on @echasnovski 's design described [here](https://github.com/neovim/neovim/issues/32537#issuecomment-2930373767).

comfysage committed:
feat(ui): progress interface

justinmk commented:
I think @shadmansaleh may have some work in-progress for this. Or if not, then we can perhaps coordinate here. Though first we need to decide what the interface and requirements are.

comfysage committed:
fix(ui): cleanup progress hook signature

comfysage commented:
> I think @shadmansaleh may have some work in-progress for this. Or if not, then we can perhaps coordinate here. Though first we need to decide what the interface and requirements are.

oh so sorry; i was not aware someone else had already started working on this.
i wanted to create an starting point to create some discussion on feedback iteration since i felt that the original issue had become sort of stale (mostly everyone agreed on the requirements and how it should be used).
i felt like it was important to get some work started since some other pr's seemed to depend on it to some capacity (#34009, #28809).

comfysage committed:
fix(ui): add `vim.ui.progress` types

comfysage committed:
docs: add `vim.ui.progress.*`

shadmansaleh commented:
> I think @shadmansaleh may have some work in-progress for this. Or if not, then we can perhaps coordinate here. Though first we need to decide what the interface and requirements are.

I actually haven't yet started work on this. Had ideas on it. But after you said it'll be part of @ofseed s project I didn't touch it.

justinmk commented:
@ofseed mentioned that the "progress" feature may involve changes to the LSP layer. whereas this PR is focused on the api. 

we can discuss the api here without overlapping the lsp-related changes.

ofseed commented:
Yes, my understanding is that I will use the progress interface to implement LSP-related features, but it does not include implementing the interface itself. I think our division of work is that my work focuses on LSP, while @shadmansaleh 's work focuses on the status line (which also includes `vim.ui.progress`) and the prompt buffer.

echasnovski commented on runtime/lua/vim/ui.lua:
```diff
@@ -235,4 +235,162 @@ function M._get_urls()
+}
+
+M._progress.on_new = function(_, _, task)
+  vim.notify(
```
Using `vim.notify()` in the default implementation will be problematic for users who have `vim.notify` implementations that show notifications outside of command line. A very common approach is to use floating windows to show each notification separately and hide it after a certain period of time.

The problem here is that using `vim.notify` will result in an overwhelming amount of separate notifications for each progress step.

In a sense, `vim.ui.progress` is intentionally meant to be a separate from `vim.notify` method to show "notifications that should be updated in place", i.e. not result in a separate notifications. In theory, if `vim.notify` was initially designed (or allowed to be updated) to emit and respect some kind of `token`/`id` for separate notifications, there might not have been a need for a `vim.ui.progress` in the first place.

So I think default `vim.ui.progress` implementation should use `print` or `vim.api.nvim_echo`.

comfysage committed:
feat(ui): use `progress.percentage` instead of `done` and `total` fields

comfysage committed:
fix(ui): progress report update logic

- `task.message` not getting updated
- `task.cancellable`: clean up if statement

comfysage committed:
feat(ui): use `nvim_echo` for progress messages

comfysage commented:
i changed the progress messages to use `vim.api.nvim_echo` instead and was wondering if it should use a custom `ui-messages` kind to indicate a progress message.

clason commented:
No, we absolutely want both. The statusline is the right place for a simple spinner or percent count, but (LSP) progress notifications can come with (arbitrarily long) messages that have no place in the statusline but may still be shown to users (e.g., in a fixed `pos = 'box'` extui place).

comfysage commented:
> If users/plugins want to create their own way of showing progress updates, all they have to do is override `vim.ui.progress`. The same approach as with `vim.ui.select`, `vim.ui.input`, and `vim.notify`.

my initial idea in this pr *was* that users could override the `on_*` functions to separate ui and core logic. the main function can handle state changes through a *static* api and then ui could be handled however someone might want.

if this is not what is expected from this feature I'll replace it. <3

comfysage commented:
> My personal preference would be to not overcomplicate this with custom Lua solutions and instead trigger something like `ProgressChanged` event with necessary data.

i feel like this is a solution that fits how other neovim interfaces have been designed.
using autocommands instead of listeners better fits the current design of other neovim api's.

comfysage committed:
feat(ui): move to autocmd based progress listeners

mfussenegger commented:
The proposal so far looks a lot like the LSP [progress support](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#progress) / [work done progress](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workDoneProgress).

What's the advantage of this, and what are the shortcomings of the LSP progress approach?

comfysage committed:
feat(ui): simplify autocmd structure

justinmk commented:
Before introducing a new Progress object, let's try to instead augment existing vim `:messages`. They can take an id, be updated, and emit events.

Updated proposal: https://github.com/neovim/neovim/issues/32537#issuecomment-2672600804
